### PR TITLE
Add fence suppot

### PR DESCRIPTION
**Pull request by csherratt**
Solves #688

This exposes the command fences to the developer. Via the `DeviceFence` trait. Fences are managed much like other resources and are added as an associated type. This api adds two abilities, the fence can block the device server by using `fenced_submit`. Or the user may synchronize in the client cpu by using  `fence_wait`.

### CODE REVIEW

File: src/backend/gl/src/lib.rs [hunk 1]
```diff
@@ -640,6 +647,35 @@ impl gfx::Device for Device {
...
+                let f = self.frame_handles.ref_fence(&fence);
+                self.share.context.WaitSync(f.0, 0, gl::TIMEOUT_IGNORED);
+            }
+        }
+
+        self.submit(info);
+
+        unsafe {
+            let fence = self.share.context.FenceSync(gl::SYNC_GPU_COMMANDS_COMPLETE, 0);
+            self.share.handles.borrow_mut().make_fence(Fence(fence))
```
Comment (kvark) [inline src/backend/gl/src/lib.rs]: Does this line need to be inside `unsafe{}` ?

File: src/core/src/device/handle.rs [hunk 2]
```diff
@@ -163,6 +163,9 @@ impl<R: Resources> Sampler<R> {
     }
 }
 
+/// Sampler Handle
```
Comment (kvark) [inline src/core/src/device/handle.rs]: Comment needed

**Comment by kvark** [comment]
Some small nitpicks, otherwise - great stuff!

File: src/backend/gl/src/lib.rs (hunk 1 above)
Comment (csherratt) [inline src/backend/gl/src/lib.rs]: the GL call is marked as unsafe, but the `make_fence` is not.

Comment (kvark) [inline src/backend/gl/src/lib.rs]: So we can do something like this instead:
```rust
let fence = Fence(unsafe {
    self.share.context.FenceSync(gl::SYNC_GPU_COMMANDS_COMPLETE, 0)
});
self.share.handles.borrow_mut().make_fence(fence)
```

**Comment by kvark** [comment]
Thanks! Are you ready to merge?

**Comment by csherratt** [comment]
@kvark I think it is good to merge.

**Comment by kvark** [comment]
@homu r+

**Pull request merged**